storage/raftstorage.go

```diff
@@ -67,80 +67,88 @@ func (r *Range) InitialState() (raftpb.HardState, raftpb.ConfState, error) {
 	return hs, cs, nil
 }
 
-// loadLastIndex looks in the engine to find the last log index.
+// setLastIndex persists a new last index and updates its cached
+// value in the range. If a non-nil engine is passed, it is used for the
+// write; otherwise the range's default engine is used. Commit() will
+// be called on the engine to keep the update as atomic as possible.
+func (r *Range) setLastIndex(eng engine.Engine, lastIndex uint64) error {
```

Comment:
This is confusing. If you really want to make sure this value is persisted immediately, then just use r.rm.Engine().

If you pass something in here which is a batch, call it batch. Otherwise, it's just too confusing. Are there multiple callers of this, some of which present a batch, others an engine? If so, you definitely don't want to commit the caller's batch. That's really bad form!